### DISCUSSION TITLE
When will the autofree be production ready?

### DISCUSSION THREAD

**Discussion by Anivie** [vlang/v#16553] [Questions and Answers] [answered]
I'm curious what V's current memory management is,but the doc describe it by a very short way.
>Most objects (~90-100%) are freed by V's autofree engine: the compiler inserts necessary free calls automatically during compilation. Remaining small percentage of objects is freed via reference counting.

So,is that ~90-100% objects will free by the autofree engine if it enable?But now this engine is not enable by default.
Then,what will v free the object now?Using the gc at all?If it,when will the autofree funtion will production ready?

**Comment by JalonSolov** [comment, marked as answer]
V does use gc by default now.  You can disable it if you like, by using `-gc none`, or you can combine it with `-autofree` to test.  `-autofree` will be ready before the 1.0 release, but there are no dates.

**Reply by Anivie** [reply]
Thanks for your answer.
This sounds very far away,from the current update frequency, we need to wait over three years?

**Reply by JalonSolov** [reply]
The current update frequency is actually quite fast.  The version numbers don't change, but there are a _lot_ of commits every week.

It's more important to get fixes done than to worry about updating the version numbers.

**Reply by Anivie** [reply]
Thanks,I just can't imagine how many functions still need to be implemented in this engine.

**Reply by Wajinn** [reply]
> Thanks for your answer. This sounds very far away,from the current update frequency, we need to wait over three years?

Is there something that you are doing, where you need autofree immediately?  The wording of the posts gives almost the impression of asking out of desperation.

For the majority of users, GC would be enough, and they don't have to worry about manual memory management (which can be a distraction or annoyance). Having an optional GC also makes the language great for casuals. This includes recently that government agencies have been recommending various GC languages to prevent common memory safety issues.  V is also a compiled language, so even using a GC, its faster than many others.  Especially considering it is not yet optimized (in comparison to vastly older languages) for many of the tests its being compared with.

V users can disable the GC, and manage memory manually. See features such as "manualfree", mentioned in the documentation.  Though perhaps it might be great if some of the developers go into greater detail in the documentation about manual memory management with V and perhaps separately post any examples, experiences, and recommendations for going that route. 

It should perhaps be made more clear to newer users and in the documentation, that V is providing 3 options for memory management. GC, autofree, and manual. This could include going into GC combined with autofree, versus separate usage. That way it might cut down on any confusion that they may have or is being put out there by competitors.

**Reply by Anivie** [reply]
> > Thanks for your answer. This sounds very far away,from the current update frequency, we need to wait over three years?
> 
> Is there something that you are doing, where you need autofree immediately? The wording of the posts gives almost the impression of asking out of desperation.
> 
> For the majority of users, GC would be enough, and they don't have to worry about manual memory management (which can be a distraction or annoyance). Having an optional GC also makes the language great for casuals. This includes recently that government agencies have been recommending various GC languages to prevent common memory safety issues. V is also a compiled language, so even using a GC, its faster than many others. Especially considering it is not yet optimized (in comparison to vastly older languages) for many of the tests its being compared with.
> 
> V users can disable the GC, and manage memory manually. See features such as "manualfree", mentioned in the documentation. Though perhaps it might be great if some of the developers go into greater detail in the documentation about manual memory management with V and perhaps separately post any examples, experiences, and recommendations for going that route.
> 
> It should perhaps be made more clear to newer users and in the documentation, that V is providing 3 options for memory management. GC, autofree, and manual. This could include going into GC combined with autofree, versus separate usage. That way it might cut down on any confusion that they may have or is being put out there by competitors.

Thanks for your explanation,I am already have some gc languages in use,and the vlang that attracts me is the memory management in compile time and the performance,so I maybe look anxious.

**Comment by Jengro777** [comment]
How should I disable gc and use autofree

`v  run -gc none vweb_translate.v`
or
`v  run -gc none -autofree vweb_translate.v`

**Reply by JalonSolov** [reply]
The 2nd.
